### PR TITLE
feat: add new service mqcloud

### PR DESCRIPTION
**Pull request by shuaibabubakker**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Introduce `ibm_mqcloud` service with these resources.
* mqcloud_queue_manager
* mqcloud_application
* mqcloud_user
* mqcloud_keystore_certificate
* mqcloud_truststore_certificate

Output from acceptance testing:

```
> make testacc TEST=./ibm/service/mqcloud/...
=== RUN   TestAccIbmMqcloudApplicationDataSourceBasic
--- PASS: TestAccIbmMqcloudApplicationDataSourceBasic (34.09s)
=== RUN   TestAccIbmMqcloudApplicationDataSourceAllArgs
--- PASS: TestAccIbmMqcloudApplicationDataSourceAllArgs (30.01s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateDataSourceBasic
--- PASS: TestAccIbmMqcloudKeystoreCertificateDataSourceBasic (44.28s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateDataSourceAllArgs
--- PASS: TestAccIbmMqcloudKeystoreCertificateDataSourceAllArgs (47.58s)
=== RUN   TestAccIbmMqcloudQueueManagerStatusDataSourceBasic
=== PAUSE TestAccIbmMqcloudQueueManagerStatusDataSourceBasic
=== RUN   TestAccIbmMqcloudQueueManagerDataSourceBasic
=== PAUSE TestAccIbmMqcloudQueueManagerDataSourceBasic
=== RUN   TestAccIbmMqcloudQueueManagerDataSourceAllArgs
=== PAUSE TestAccIbmMqcloudQueueManagerDataSourceAllArgs
=== RUN   TestAccIbmMqcloudTruststoreCertificateDataSourceBasic
--- PASS: TestAccIbmMqcloudTruststoreCertificateDataSourceBasic (43.26s)
=== RUN   TestAccIbmMqcloudTruststoreCertificateDataSourceAllArgs
--- PASS: TestAccIbmMqcloudTruststoreCertificateDataSourceAllArgs (43.42s)
=== RUN   TestAccIbmMqcloudUserDataSourceBasic
--- PASS: TestAccIbmMqcloudUserDataSourceBasic (35.73s)
=== RUN   TestAccIbmMqcloudUserDataSourceAllArgs
--- PASS: TestAccIbmMqcloudUserDataSourceAllArgs (38.38s)
=== RUN   TestAccIbmMqcloudApplicationBasic
--- PASS: TestAccIbmMqcloudApplicationBasic (35.67s)
=== RUN   TestAccIbmMqcloudApplicationAllArgs
--- PASS: TestAccIbmMqcloudApplicationAllArgs (38.85s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateBasic
--- PASS: TestAccIbmMqcloudKeystoreCertificateBasic (47.55s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateAllArgs
--- PASS: TestAccIbmMqcloudKeystoreCertificateAllArgs (51.08s)
=== RUN   TestAccIbmMqcloudQueueManagerBasic
=== PAUSE TestAccIbmMqcloudQueueManagerBasic
=== RUN   TestAccIbmMqcloudQueueManagerAllArgs
=== PAUSE TestAccIbmMqcloudQueueManagerAllArgs
=== RUN   TestAccIbmMqcloudTruststoreCertificateBasic
--- PASS: TestAccIbmMqcloudTruststoreCertificateBasic (47.99s)
=== RUN   TestAccIbmMqcloudTruststoreCertificateAllArgs
--- PASS: TestAccIbmMqcloudTruststoreCertificateAllArgs (48.27s)
=== RUN   TestAccIbmMqcloudUserBasic
--- PASS: TestAccIbmMqcloudUserBasic (26.36s)
=== RUN   TestAccIbmMqcloudUserAllArgs
--- PASS: TestAccIbmMqcloudUserAllArgs (31.26s)
=== CONT  TestAccIbmMqcloudQueueManagerStatusDataSourceBasic
=== CONT  TestAccIbmMqcloudQueueManagerBasic
=== CONT  TestAccIbmMqcloudQueueManagerDataSourceAllArgs
=== CONT  TestAccIbmMqcloudQueueManagerDataSourceBasic
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
--- PASS: TestAccIbmMqcloudQueueManagerStatusDataSourceBasic (27.12s)
=== CONT  TestAccIbmMqcloudQueueManagerAllArgs
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the initializing state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the running state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the deploying state ....
The queue manager is currently in the running state ....
The queue manager is currently in the running state ....
The queue manager is currently in the running state ....
--- PASS: TestAccIbmMqcloudQueueManagerDataSourceBasic (287.96s)
The queue manager is currently in the upgrading_version state ....
--- PASS: TestAccIbmMqcloudQueueManagerBasic (319.60s)
--- PASS: TestAccIbmMqcloudQueueManagerDataSourceAllArgs (320.02s)
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the upgrading_version state ....
The queue manager is currently in the running state ....
--- PASS: TestAccIbmMqcloudQueueManagerAllArgs (598.01s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/mqcloud 1270.072s
```

### CODE REVIEW

File: ibm/conns/config.go [hunk 1]
```diff
@@ -87,6 +87,7 @@ import (
 	"github.com/apache/openwhisk-client-go/whisk"
 	jwt "github.com/golang-jwt/jwt"
 	slsession "github.com/softlayer/softlayer-go/session"
+	"github.ibm.com/mqoc/mqcloud-go-sdk/mqcloudv1"
```
[the remark this answers is not in the document]
Comment (shuaibabubakker) [reply ibm/conns/config.go:90]: sure

File: ibm/conns/config.go [hunk 2]
```diff
@@ -3258,6 +3269,30 @@ func (c *Config) ClientSession() (interface{}, error) {
 		URL:           EnvFallBack([]string{"IBMCLOUD_CODE_ENGINE_API_ENDPOINT"}, codeEngineEndpoint),
 	}
 
+	// MQ Cloud Service Configuration
+	mqCloudConfigURL := mqcloudv1.DefaultServiceURL
```
[the remark this answers is not in the document]
Comment (shuaibabubakker) [reply ibm/conns/config.go:3273]: yeah its a region specific endpoint.

```
// DefaultServiceURL is the default URL to make service requests to.
const DefaultServiceURL = "https://api.private.eu-fr2.mq2.cloud.ibm.com"

// DefaultServiceName is the default key used to find external configuration information.
const DefaultServiceName = "mqcloud"

const ParameterizedServiceURL = "https://api.private.{region}.mq2.cloud.ibm.com"

var defaultUrlVariables = map[string]string{
	"region": "eu-fr2",
}
```
So we should consider setting the url something like this 

`ContructEndpoint(fmt.Sprintf("api.private.%s.mq2", c.Region), cloudEndpoint)`

and we should consider the visibilty cases too. right?

Got it. Will do that.

File: ibm/service/mqcloud/data_source_ibm_mqcloud_queue_manager.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The GUID that uniquely identifies the MQ on Cloud service instance.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "A queue manager name conforming to MQ restrictions.",
+			},
+			"first": {
```
[the remark this answers is not in the document]
Comment (shuaibabubakker) [reply ibm/service/mqcloud/data_source_ibm_mqcloud_queue_manager.go:34]: Ok will try that one.

File: ibm/service/mqcloud/resource_ibm_mqcloud_queue_manager.go [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+	setQueueManagerVersionOptions.SetServiceInstanceGuid(parts[0])
+	setQueueManagerVersionOptions.SetQueueManagerID(parts[1])
+
+	hasChange := false
+
+	if d.HasChange("service_instance_guid") {
+		return diag.FromErr(fmt.Errorf("Cannot update resource property \"%s\" with the ForceNew annotation."+
+			" The resource must be re-created to update this property.", "service_instance_guid"))
+	}
+	if d.HasChange("name") {
```
[the remark this answers is not in the document]
Comment (shuaibabubakker) [reply ibm/service/mqcloud/resource_ibm_mqcloud_queue_manager.go:355]: ok

File: ibm/service/mqcloud/utils.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+			queueManagerDetails, response, err := mqcloudClient.GetQueueManagerWithContext(context, getQueueManagerOptions)
+			if err != nil {
+				if response != nil && response.StatusCode == 404 {
+					return queueManagerDetails, isQueueManagerDeleteDone, nil
+				}
+				return nil, "", err
+			}
+			return queueManagerDetails, isQueueManagerDeleting, nil
+		},
+		Timeout:    10 * time.Minute,
```
[the remark this answers is not in the document]
Comment (shuaibabubakker) [reply ibm/service/mqcloud/utils.go:100]: yeah got it

File: ibm/conns/config.go [hunk 6]
```diff
@@ -1214,6 +1219,13 @@ func (session clientSession) ProjectV1() (*project.ProjectV1, error) {
 	return session.projectClient, session.projectClientErr
 }
 
+// MQ on Cloud
+func (session clientSession) MqcloudV1() (*mqcloudv1.MqcloudV1, error) {
+	sessionMqcloudClient := session.mqcloudClient
+	sessionMqcloudClient.EnableRetries(0, 0)
```
Comment (shuaibabubakker) [inline ibm/conns/config.go:1225]: We have added default retry mechanism here.
Need a review here, if its fine to do.

File: ibm/service/mqcloud/utils.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+			getQueueManagerStatusOptions.SetServiceInstanceGuid(parts[0])
+			getQueueManagerStatusOptions.SetQueueManagerID(parts[1])
+			queueManagerStatus, response, err := mqcloudClient.GetQueueManagerStatusWithContext(context, getQueueManagerStatusOptions)
+			if err != nil {
+				return "", "", fmt.Errorf("GetQueueManagerWithContext ...... %s err: %s", response, err)
+			}
+			if queueManagerStatus == nil || queueManagerStatus.Status == nil {
+				return nil, "", fmt.Errorf("queueManagerStatus or queueManagerStatus.Status is nil")
+			}
+			fmt.Println("The queue manager is currently in the " + *queueManagerStatus.Status + " state ....")
```
Comment (hkantare) [inline ibm/service/mqcloud/utils.go:61]: Can we remove this fmt stmt i think its not required Terraform will log it still creating...

File: website/allowed-subcategories.txt [hunk 8]
```diff
@@ -27,6 +27,7 @@ Key Management Service
 Kubernetes Service
 Logs Router
 Metrics Router
+MQ on Cloud
```
Comment (hkantare) [inline website/allowed-subcategories.txt:30]: Is this same name refered generally on your cli,api ,cloud docs?
Do we need to mention explictly "on Cloud"

**Comment by hkantare** [comment]
Resolve conflicts on above files

File: website/allowed-subcategories.txt (hunk 8 above)
Comment (shuaibabubakker) [reply website/allowed-subcategories.txt:30]: Yes, we need to explicitly mention 'on Cloud' here, as both MQ and MQ on Cloud are separate services.

In the apidocs and sdk we are following this name itself 
- api file: https://github.ibm.com/cloud-api-docs/mq-on-cloud/blob/6840e3fcdfea5cf6ebf9f2a742d3fc98da288cca/mq-on-cloud.yaml#L4
- sdk: https://github.com/IBM/mqcloud-go-sdk/#ibm-cloud-mqoc-services-go-sdk-001

**Comment by shuaibabubakker** [comment]
Ran the acc test once again.

```
=== RUN   TestAccIbmMqcloudApplicationDataSourceBasic
--- PASS: TestAccIbmMqcloudApplicationDataSourceBasic (53.41s)
=== RUN   TestAccIbmMqcloudApplicationDataSourceAllArgs
--- PASS: TestAccIbmMqcloudApplicationDataSourceAllArgs (41.24s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateDataSourceBasic
--- PASS: TestAccIbmMqcloudKeystoreCertificateDataSourceBasic (55.50s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateDataSourceAllArgs
--- PASS: TestAccIbmMqcloudKeystoreCertificateDataSourceAllArgs (53.99s)
=== RUN   TestAccIbmMqcloudQueueManagerStatusDataSourceBasic
=== PAUSE TestAccIbmMqcloudQueueManagerStatusDataSourceBasic
=== RUN   TestAccIbmMqcloudQueueManagerDataSourceBasic
=== PAUSE TestAccIbmMqcloudQueueManagerDataSourceBasic
=== RUN   TestAccIbmMqcloudQueueManagerDataSourceAllArgs
=== PAUSE TestAccIbmMqcloudQueueManagerDataSourceAllArgs
=== RUN   TestAccIbmMqcloudTruststoreCertificateDataSourceBasic
--- PASS: TestAccIbmMqcloudTruststoreCertificateDataSourceBasic (49.71s)
=== RUN   TestAccIbmMqcloudTruststoreCertificateDataSourceAllArgs
--- PASS: TestAccIbmMqcloudTruststoreCertificateDataSourceAllArgs (43.17s)
=== RUN   TestAccIbmMqcloudUserDataSourceBasic
--- PASS: TestAccIbmMqcloudUserDataSourceBasic (41.22s)
=== RUN   TestAccIbmMqcloudUserDataSourceAllArgs
--- PASS: TestAccIbmMqcloudUserDataSourceAllArgs (39.58s)
=== RUN   TestAccIbmMqcloudApplicationBasic
--- PASS: TestAccIbmMqcloudApplicationBasic (35.25s)
=== RUN   TestAccIbmMqcloudApplicationAllArgs
--- PASS: TestAccIbmMqcloudApplicationAllArgs (43.11s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateBasic
--- PASS: TestAccIbmMqcloudKeystoreCertificateBasic (48.02s)
=== RUN   TestAccIbmMqcloudKeystoreCertificateAllArgs
--- PASS: TestAccIbmMqcloudKeystoreCertificateAllArgs (51.40s)
=== RUN   TestAccIbmMqcloudQueueManagerBasic
=== PAUSE TestAccIbmMqcloudQueueManagerBasic
=== RUN   TestAccIbmMqcloudQueueManagerAllArgs
=== PAUSE TestAccIbmMqcloudQueueManagerAllArgs
=== RUN   TestAccIbmMqcloudTruststoreCertificateBasic
--- PASS: TestAccIbmMqcloudTruststoreCertificateBasic (39.13s)
=== RUN   TestAccIbmMqcloudTruststoreCertificateAllArgs
--- PASS: TestAccIbmMqcloudTruststoreCertificateAllArgs (46.67s)
=== RUN   TestAccIbmMqcloudUserBasic
--- PASS: TestAccIbmMqcloudUserBasic (39.19s)
=== RUN   TestAccIbmMqcloudUserAllArgs
--- PASS: TestAccIbmMqcloudUserAllArgs (43.04s)
=== CONT  TestAccIbmMqcloudQueueManagerBasic
=== CONT  TestAccIbmMqcloudQueueManagerStatusDataSourceBasic
=== CONT  TestAccIbmMqcloudQueueManagerDataSourceAllArgs
=== CONT  TestAccIbmMqcloudQueueManagerDataSourceBasic
--- PASS: TestAccIbmMqcloudQueueManagerStatusDataSourceBasic (34.88s)
=== CONT  TestAccIbmMqcloudQueueManagerAllArgs
--- PASS: TestAccIbmMqcloudQueueManagerDataSourceBasic (322.30s)
--- PASS: TestAccIbmMqcloudQueueManagerBasic (364.36s)
--- PASS: TestAccIbmMqcloudQueueManagerDataSourceAllArgs (383.86s)
--- PASS: TestAccIbmMqcloudQueueManagerAllArgs (629.05s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/mqcloud 1366.661s

```

File: website/allowed-subcategories.txt (hunk 8 above)
Comment (shuaibabubakker) [reply website/allowed-subcategories.txt:29]: Hey @hkantare 

After discussing with the team about the name we should use, we encountered some conflicts regarding the product name. Here is the full list of names we have been using:

1. `MQ` - used in the IBM Cloud catalog (UI)
2. `mqcloud` - registered service name in RMC
3. `MQ on Cloud` - used in the doc and APIs, SDKs
4. `IBM MQ on Cloud` - used in docs
5. `IBM MQ on IBM Cloud` - Official registered name of current service
6. `IBM MQ on IBM Cloud Reserved Instance` - Official name of BNPP offering

So which name would you suggest us to use?

Comment (shuaibabubakker) [reply website/allowed-subcategories.txt:29]: After some more discussions. we are right now sticking with `MQ on Cloud` itself 👍

**Pull request merged**